cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,466 @@
+The mTLS handlers will also need to indicate the remote node, along with the auth type.
+For that lifetime, traffic between Identity A and Identity B will flow without further checks.
+
+#### SPIFFE installation steps and flow
+
+1. User creates a cluster, without CNI, per usual Cilium install process. 
+1. User installs Cilium, with mTLS enabled, SPIFFE auth will not work until the following conditions are met:
+    * SPIRE Server must be installed somewhere and configured correctly. (Bring-your-own-SPIRE-Server is an anticipated future need, but the initial implementation will use an in-cluster SPIRE server.)
+    * Per-node SPIRE Agent must be installed in the cluster and configured correctly. The SPIRE agent talks to the SPIRE server over the network, but all other communication with the Cilium Agent is via domain sockets shared on the host's filesystem.
+    * Cilium Agent SPIFFE Identity must be created as per SPIRE Agent config.
+    * Cilium Operator SPIFFE Identity must be created as per SPIRE Server config.
+    * (This document will be updated after [#23806](https://github.com/cilium/cilium/issues/23806) is done.)
+1. Cilium agent starts up as normal, acts as CNI.
+1. Cilium agent also contacts the local SPIRE agent at startup (via a domain socket shared on the host's filesystem) to watch the Delegated Identity API and gets its own SPIFFE identity via the SPIRE workload API.
+1. When generating Cilium Security Identities for identities with mTLS auth enabled, Cilium Operator in SPIFFE mode also records the SPIFFE identity (that is, the string `spiffe://spiffe.cilium/identity/1337`).
+For SPIRE implementations, the Cilium Operator also creates the required SPIRE server entries.
+1. When checking auth-enabled policy, the datapath will flag that there is no entry in the auth table for the flow, and then the agent will use its cache of keypairs to handshakes with the agent on the other end of the connection (either source node agent will reach out to destination node agent, or vice versa), and performs a TLS handshake. Note that an optimization which may be included here is that the agent may pre-authenticate identities that it knows will need it, if we can make that work.
+1. If the TLS handshake succeeds, then the connection is authenticated and the traffic protected by policy can proceed. The TLS connection between the agents can then be torn down and thrown away (since it’s not being used to tunnel the encrypted traffic through).
+1. The details of the particular flow are saved to the auth table by the agent; note that this needs to include the remote node the auth was performed with.
+It's not possible to reuse an identity-identity handshake across nodes without making bad-node attacks much easier.
+
+
+### Control Plane
+
+#### Connections
+
+##### Between Cilium Operator instances and the SPIFFE server
```

Comment:
@youngnick not sure what your plans are regarding this CFP, but I noted that the diagrams in the CFP don't include the operator in the architecture. If you plan to update the CFP with the final design, it would be helpful for readers to see this interaction in one of the diagrams.

Reply:
I have been hoping to come back and do a final update pass, yes. Thanks for the tip.